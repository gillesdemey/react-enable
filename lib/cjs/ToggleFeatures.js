"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToggleFeatures = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var FeatureContext_1 = require("./FeatureContext");
var ToggleFeatures = function () {
    var _a = react_1.default.useState(false), showDrawer = _a[0], setShowDrawer = _a[1];
    var context = react_1.default.useContext(FeatureContext_1.FeatureContext);
    if (context == null) {
        return null;
    }
    var dispatch = context.dispatch, state = context.state;
    if (state.features.length === 0) {
        return null;
    }
    return (react_1.default.createElement("aside", { className: "feature-flags", style: {
            left: showDrawer ? "0" : "-198px",
            top: "100px",
            width: "220px",
            height: "300px",
            color: "black",
            background: "white",
            border: "1px solid gray",
            zIndex: 100000,
            opacity: showDrawer ? "1" : "0.6",
            position: "fixed",
            display: "flex",
            flexDirection: "row",
            justifyContent: "space-between",
            flexWrap: "nowrap",
            alignItems: "stretch",
            alignContent: "stretch"
        } },
        react_1.default.createElement("main", { style: {
                flexGrow: 1,
                padding: "12px",
                overflow: "scroll"
            } }, state.features.map(function (feature) { return (react_1.default.createElement("div", { key: feature.name },
            react_1.default.createElement("label", null,
                react_1.default.createElement("input", { type: "checkbox", onChange: function () {
                        dispatch({ type: "toggle", feature: feature.name });
                    }, checked: state.currentEnabled.has(feature.name) }),
                feature.name),
            feature.description != null ? (react_1.default.createElement("p", { style: {
                    marginTop: "0",
                    marginLeft: "1.5em",
                    fontSize: "0.5rem",
                    lineHeight: "0.5rem"
                } }, feature.description)) : null)); })),
        react_1.default.createElement("div", { style: {
                background: "rgb(238, 238, 238)",
                padding: "10px 0",
                writingMode: "vertical-rl",
                flexGrow: 0,
                width: "24px",
                display: "flex",
                flexDirection: "row",
                cursor: "pointer",
                justifyContent: "space-between",
                alignItems: "center"
            }, onClick: function () { return setShowDrawer(!showDrawer); } },
            react_1.default.createElement("button", { style: {
                    border: "none",
                    fontSize: "8px",
                    background: "transparent",
                    padding: "2px",
                    cursor: "pointer"
                } }, showDrawer ? "ðŸŽ› Hide" : "ðŸŽ› Show"),
            react_1.default.createElement("span", null, "Dev Features Panel"))));
};
exports.ToggleFeatures = ToggleFeatures;
