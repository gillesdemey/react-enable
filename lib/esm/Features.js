import * as React from "react";
import { EnableContext } from "./EnableContext";
import { fromStrArray, reducer, elemStr } from "./FeatureState";
import { GlobalEnable } from "./GlobalEnable";
import { FeatureContext } from "./FeatureContext";
export var Features = function (_a) {
    var features = _a.features, defaultEnabled = _a.defaultEnabled, _b = _a.consoleOverride, consoleOverride = _b === void 0 ? false : _b, storage = _a.storage, children = _a.children;
    var initial = {
        features: features,
        forceDisable: fromStrArray([]),
        forceEnable: fromStrArray([]),
        defaultActive: fromStrArray(defaultEnabled),
        currentEnabled: new Set()
    };
    var _c = React.useReducer(reducer, initial), state = _c[0], dispatch = _c[1];
    useLoad(storage, dispatch);
    usePersist(state.forceEnable, state.forceDisable, storage);
    useConsoleOverride(consoleOverride, dispatch);
    React.useEffect(function () {
        if (Array.isArray(defaultEnabled)) {
            dispatch({ type: "set-default", defaultEnabled: defaultEnabled });
        }
    }, [dispatch, defaultEnabled]);
    var featureValue = React.useMemo(function () { return ({ dispatch: dispatch, state: state }); }, [
        dispatch,
        state,
        state.defaultActive,
        state.forceEnable,
        state.forceDisable
    ]);
    var testCallback = React.useCallback(function (f) { return elemStr(f, state.currentEnabled); }, [state.currentEnabled]);
    return (React.createElement(FeatureContext.Provider, { value: featureValue },
        React.createElement(EnableContext.Provider, { value: testCallback }, children)));
};
function useConsoleOverride(consoleOverride, dispatch) {
    React.useEffect(function () {
        if (!consoleOverride) {
            return function () {
            };
        }
        window.feature = new GlobalEnable(dispatch);
        return function () {
            window.feature = undefined;
        };
    }, [dispatch]);
}
function useLoad(storage, dispatch) {
    React.useEffect(function () {
        if (storage != null) {
            try {
                var value = storage.getItem("react-enable:features");
                if (value != null) {
                    var result = JSON.parse(value);
                    if (Array.isArray(result.on) &&
                        Array.isArray(result.off) &&
                        typeof result.on[0] === "string" &&
                        typeof result.off[0] === "string") {
                        dispatch({ type: "set-active", values: result });
                    }
                }
            }
            catch (e) {
            }
        }
    }, [storage]);
}
function usePersist(forceEnable, forceDisable, storage) {
    var strState = storage == null
        ? undefined
        : JSON.stringify({
            on: Array.from(forceEnable).sort(),
            off: Array.from(forceDisable).sort()
        });
    React.useEffect(function () {
        if (strState != null && storage != null) {
            try {
                storage.setItem("react-enable:features", strState);
            }
            catch (e) {
            }
        }
    }, [storage, strState]);
}
