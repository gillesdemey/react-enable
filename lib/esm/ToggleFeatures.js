import React from "react";
import { FeatureContext } from "./FeatureContext";
export var ToggleFeatures = function () {
    var _a = React.useState(false), showDrawer = _a[0], setShowDrawer = _a[1];
    var context = React.useContext(FeatureContext);
    if (context == null) {
        return null;
    }
    var dispatch = context.dispatch, state = context.state;
    if (state.features.length === 0) {
        return null;
    }
    return (React.createElement("aside", { className: "feature-flags", style: {
            left: showDrawer ? "0" : "-198px",
            top: "100px",
            width: "220px",
            height: "300px",
            color: "black",
            background: "white",
            border: "1px solid gray",
            zIndex: 100000,
            opacity: showDrawer ? "1" : "0.6",
            position: "fixed",
            display: "flex",
            flexDirection: "row",
            justifyContent: "space-between",
            flexWrap: "nowrap",
            alignItems: "stretch",
            alignContent: "stretch"
        } },
        React.createElement("main", { style: {
                flexGrow: 1,
                padding: "12px",
                overflow: "scroll"
            } }, state.features.map(function (feature) { return (React.createElement("div", { key: feature.name },
            React.createElement("label", null,
                React.createElement("input", { type: "checkbox", onChange: function () {
                        dispatch({ type: "toggle", feature: feature.name });
                    }, checked: state.currentEnabled.has(feature.name) }),
                feature.name),
            feature.description != null ? (React.createElement("p", { style: {
                    marginTop: "0",
                    marginLeft: "1.5em",
                    fontSize: "0.5rem",
                    lineHeight: "0.5rem"
                } }, feature.description)) : null)); })),
        React.createElement("div", { style: {
                background: "rgb(238, 238, 238)",
                padding: "10px 0",
                writingMode: "vertical-rl",
                flexGrow: 0,
                width: "24px",
                display: "flex",
                flexDirection: "row",
                cursor: "pointer",
                justifyContent: "space-between",
                alignItems: "center"
            }, onClick: function () { return setShowDrawer(!showDrawer); } },
            React.createElement("button", { style: {
                    border: "none",
                    fontSize: "8px",
                    background: "transparent",
                    padding: "2px",
                    cursor: "pointer"
                } }, showDrawer ? "ðŸŽ› Hide" : "ðŸŽ› Show"),
            React.createElement("span", null, "Dev Features Panel"))));
};
