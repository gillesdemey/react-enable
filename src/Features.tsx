import * as React from "react";
import { EnableContext } from "./EnableContext";
import {
  Feature,
  DispatchFeature,
  FeatureState,
  fromStrArray,
  reducer,
  elemStr
} from "./FeatureState";
import { GlobalEnable } from "./GlobalEnable";
import { FeatureContext } from "./FeatureContext";

interface FeatureProps {
  readonly features: Feature[];
  readonly defaultEnabled: string[];
  readonly consoleOverride?: boolean;
  readonly storage?: Storage;
}

/**
 * A more batteries-enabled parent component that keeps track of feature state
 * internally, and creates window.feature.enable("f") and window.feature.disable("f").
 */
export const Features: React.FC<FeatureProps> = ({
  features,
  defaultEnabled,
  consoleOverride = false,
  storage,
  children
}) => {
  const initial: FeatureState = {
    features,
    forceDisable: fromStrArray([]),
    forceEnable: fromStrArray([]),
    defaultActive: fromStrArray(defaultEnabled),
    currentEnabled: new Set()
  };

  const [state, dispatch] = React.useReducer(reducer, initial);
  // console.debug("State: \n##> %o\n##> %o\n##> %o", state.forceEnable, state.forceDisable, state.defaultActive)

  useLoad(storage, dispatch);
  usePersist(state.forceEnable, state.forceDisable, storage);
  useConsoleOverride(consoleOverride, dispatch);

  React.useEffect(() => {
    if (Array.isArray(defaultEnabled)) {
      dispatch({ type: "set-default", defaultEnabled });
    }
  }, [dispatch, defaultEnabled]);

  const featureValue = React.useMemo(() => ({ dispatch, state }), [
    dispatch,
    state,
    state.defaultActive,
    state.forceEnable,
    state.forceDisable
  ]);

  const testCallback = React.useCallback(
    (f: string) => elemStr(f, state.currentEnabled),
    [state.currentEnabled]
  );

  return (
    <FeatureContext.Provider value={featureValue}>
      <EnableContext.Provider value={testCallback}>
        {children}
      </EnableContext.Provider>
    </FeatureContext.Provider>
  );
};

function useConsoleOverride(
  consoleOverride: boolean,
  dispatch: DispatchFeature
) {
  React.useEffect(() => {
    if (!consoleOverride) {
      return () => {
        /* empty */
      };
    }
    window.feature = new GlobalEnable(dispatch);
    return () => {
      window.feature = undefined;
    };
  }, [dispatch]);
}

function useLoad(storage: Storage | undefined, dispatch: DispatchFeature) {
  React.useEffect(() => {
    if (storage != null) {
      try {
        const value = storage.getItem("react-enable:features");
        if (value != null) {
          const result = JSON.parse(value);
          if (
            Array.isArray(result.on) &&
            Array.isArray(result.off) &&
            typeof result.on[0] === "string" &&
            typeof result.off[0] === "string"
          ) {
            dispatch({ type: "set-active", values: result });
          }
        }
      } catch (e) {
        // Can't parse or get
      }
    }
  }, [storage]);
}

function usePersist(
  forceEnable: Set<string>,
  forceDisable: Set<string>,
  storage: Storage | undefined
) {
  const strState =
    storage == null
      ? undefined
      : JSON.stringify({
          on: Array.from(forceEnable).sort(),
          off: Array.from(forceDisable).sort()
        });

  React.useEffect(() => {
    if (strState != null && storage != null) {
      try {
        storage.setItem("react-enable:features", strState);
      } catch (e) {
        // Can't set for some reason
      }
    }
  }, [storage, strState]);
}
