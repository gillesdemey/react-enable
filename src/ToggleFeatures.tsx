import React from 'react';
import { FeatureContext } from './FeatureContext';
import { valueOfFeature } from './FeaturesState';

/// Permit users to override feature flags via a GUI.
/// Renders a small floating button in lower left or right, pressing it brings up
/// a list of features to toggle and their current override state. you can override on or override off,
/// or unset the override and go back to default value.
export function ToggleFeatures({}): JSX.Element | null {
  const [showDrawer, setShowDrawer] = React.useState(false);
  const context = React.useContext(FeatureContext);

  if (context == null) {
    return null;
  }

  // We want: Real value after all nestings, value of the override. we toggle override
  const { send, featuresDescription, overridesState, defaultsState, test } = context;

  if (featuresDescription.length === 0) {
    return null;
  }
  return (
    <aside>
      <main>
        {featuresDescription.map((feature) => (
          <div key={feature.name}>
            <label>
              <input
                type="checkbox"
                onChange={() => {
                  send({ type: 'TOGGLE', name: feature.name });
                }}
                checked={valueOfFeature(overridesState, feature.name)[0]}
              />
              {feature.name}
            </label>
            {feature.description != null ? <p>{feature.description}</p> : null}
          </div>
        ))}
      </main>
      <div onClick={() => setShowDrawer((old) => !old)}>
        <button
          style={{
            border: 'none',
            fontSize: '8px',
            background: 'transparent',
            padding: '2px',
            cursor: 'pointer',
          }}
        >
          {showDrawer ? 'ðŸŽ› Hide' : 'ðŸŽ› Show'}
        </button>
      </div>
    </aside>
  );
}
