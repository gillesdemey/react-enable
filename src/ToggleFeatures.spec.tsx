// import * as React from "react";
import renderer from "react-test-renderer";

import { ToggleFeatures } from "./ToggleFeatures";
import { Features } from "./Features";

it("renders ToggleFeatures with correct checkboxes checked", () => {
  const component1 = renderer.create(
    <Features defaultEnabled={[]} features={[{ name: "A" }, { name: "B" }]}>
      <ToggleFeatures />
    </Features>
  );

  const component2 = renderer.create(
    <Features defaultEnabled={["A"]} features={[{ name: "A" }, { name: "B" }]}>
      <ToggleFeatures />
    </Features>
  );

  const component3 = renderer.create(
    <Features defaultEnabled={["C"]} features={[{ name: "A" }, { name: "B" }]}>
      <ToggleFeatures />
    </Features>
  );

  expect(component1.toJSON()).toMatchSnapshot();
  expect(component2.toJSON()).toMatchSnapshot();
  expect(component3.toJSON()).toMatchSnapshot();
});
