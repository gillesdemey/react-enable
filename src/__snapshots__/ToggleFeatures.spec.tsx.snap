// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ToggleFeatures with correct checkboxes checked 1`] = `
<aside
  className="feature-flags"
  style={
    Object {
      "alignContent": "stretch",
      "alignItems": "stretch",
      "background": "white",
      "border": "1px solid gray",
      "color": "black",
      "display": "flex",
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": "300px",
      "justifyContent": "space-between",
      "left": "-198px",
      "opacity": "0.6",
      "position": "fixed",
      "top": "100px",
      "width": "220px",
      "zIndex": 100000,
    }
  }
>
  <main
    style={
      Object {
        "flexGrow": 1,
        "overflow": "scroll",
        "padding": "12px",
      }
    }
  >
    <div>
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        A
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        B
      </label>
    </div>
  </main>
  <div
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "background": "rgb(238, 238, 238)",
        "cursor": "pointer",
        "display": "flex",
        "flexDirection": "row",
        "flexGrow": 0,
        "justifyContent": "space-between",
        "padding": "10px 0",
        "width": "24px",
        "writingMode": "vertical-rl",
      }
    }
  >
    <button
      style={
        Object {
          "background": "transparent",
          "border": "none",
          "cursor": "pointer",
          "fontSize": "8px",
          "padding": "2px",
        }
      }
    >
      ðŸŽ› Show
    </button>
    <span>
      Dev Features Panel
    </span>
  </div>
</aside>
`;

exports[`renders ToggleFeatures with correct checkboxes checked 2`] = `
<aside
  className="feature-flags"
  style={
    Object {
      "alignContent": "stretch",
      "alignItems": "stretch",
      "background": "white",
      "border": "1px solid gray",
      "color": "black",
      "display": "flex",
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": "300px",
      "justifyContent": "space-between",
      "left": "-198px",
      "opacity": "0.6",
      "position": "fixed",
      "top": "100px",
      "width": "220px",
      "zIndex": 100000,
    }
  }
>
  <main
    style={
      Object {
        "flexGrow": 1,
        "overflow": "scroll",
        "padding": "12px",
      }
    }
  >
    <div>
      <label>
        <input
          checked={true}
          onChange={[Function]}
          type="checkbox"
        />
        A
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        B
      </label>
    </div>
  </main>
  <div
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "background": "rgb(238, 238, 238)",
        "cursor": "pointer",
        "display": "flex",
        "flexDirection": "row",
        "flexGrow": 0,
        "justifyContent": "space-between",
        "padding": "10px 0",
        "width": "24px",
        "writingMode": "vertical-rl",
      }
    }
  >
    <button
      style={
        Object {
          "background": "transparent",
          "border": "none",
          "cursor": "pointer",
          "fontSize": "8px",
          "padding": "2px",
        }
      }
    >
      ðŸŽ› Show
    </button>
    <span>
      Dev Features Panel
    </span>
  </div>
</aside>
`;

exports[`renders ToggleFeatures with correct checkboxes checked 3`] = `
<aside
  className="feature-flags"
  style={
    Object {
      "alignContent": "stretch",
      "alignItems": "stretch",
      "background": "white",
      "border": "1px solid gray",
      "color": "black",
      "display": "flex",
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": "300px",
      "justifyContent": "space-between",
      "left": "-198px",
      "opacity": "0.6",
      "position": "fixed",
      "top": "100px",
      "width": "220px",
      "zIndex": 100000,
    }
  }
>
  <main
    style={
      Object {
        "flexGrow": 1,
        "overflow": "scroll",
        "padding": "12px",
      }
    }
  >
    <div>
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        A
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        B
      </label>
    </div>
  </main>
  <div
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "background": "rgb(238, 238, 238)",
        "cursor": "pointer",
        "display": "flex",
        "flexDirection": "row",
        "flexGrow": 0,
        "justifyContent": "space-between",
        "padding": "10px 0",
        "width": "24px",
        "writingMode": "vertical-rl",
      }
    }
  >
    <button
      style={
        Object {
          "background": "transparent",
          "border": "none",
          "cursor": "pointer",
          "fontSize": "8px",
          "padding": "2px",
        }
      }
    >
      ðŸŽ› Show
    </button>
    <span>
      Dev Features Panel
    </span>
  </div>
</aside>
`;
